### ISSUE TITLE
To what extent do we allow invalid models to be created?

### ISSUE THREAD

**Issue by kerimoyle** [cellml/libcellml#398]
This relates to some existing issues (found while I was figuring out stuff for #350) but also relates to how #368 will play out...

To what extent can invalid models be created through the API/parser?  There are a few places where we silently prevent things which are not valid from happening.  Aren't we wanting to allow everything to be parsed/created and then validate only on demand?  For example:

```
TEST(Validator, encapsulationHierarchyNoCycles)
{
    libcellml::Validator v;
    libcellml::ModelPtr m = std::make_shared<libcellml::Model>();
    libcellml::ComponentPtr comp1 = std::make_shared<libcellml::Component>();
    libcellml::ComponentPtr comp2 = std::make_shared<libcellml::Component>();
    libcellml::ComponentPtr comp3 = std::make_shared<libcellml::Component>();

    libcellml::VariablePtr v1 = std::make_shared<libcellml::Variable>();
    libcellml::VariablePtr v2 = std::make_shared<libcellml::Variable>();
    libcellml::VariablePtr v3 = std::make_shared<libcellml::Variable>();

    m->setName("modelName");
    comp1->setName("component1");
    comp2->setName("component2");
    comp3->setName("component3");

    v1->setName("variable1");
    v2->setName("variable2");
    v3->setName("variable3");

    v1->setUnits("dimensionless");
    v2->setUnits("dimensionless");
    v3->setUnits("dimensionless");

    comp1->addVariable(v1);
    comp2->addVariable(v2);
    comp3->addVariable(v3);

    // comp1( comp2 ( comp3 )) -> valid
    m->addComponent(comp1);
    comp1->addComponent(comp2);
    comp2->addComponent(comp3);

    v.validateModel(m);
    EXPECT_EQ(size_t(0), v.errorCount());

    // Breaking the encapsulation hierarchy by creating a loop
    comp3->addComponent(comp1);
    // ... but this is prevented inside the addComponent function - fails silently - and model is unchanged
    v.validateModel(m);
    EXPECT_EQ(size_t(0), v.errorCount());
}
```

`comp1` is not added to `comp3` because the `Component::addComponent` function tests upstream parents, and does not add it if it's already in the parent list.  There are other examples in the parser (which silently skips things it doesn't understand, like unexpected strings in numbers, more than one encapsulation etc).

Is it better to create warnings where this happens (in which case it's more related to #368) or to allow the 'mistakes' to happen and only alert the user at validation?  Either way I don't think it's good to just silently not do what the user has asked for...?

**Comment by hsorby** [comment]
I agree it is not good to silently not do stuff.  I think in this particular situation it should add the component and then validation should reveal the error.

**Comment by hsorby** [comment]
I don't think the add component function stops this does it?

**Comment by kerimoyle** [comment]
The test above gives no errors, and the addComponent function does not add
it because it calls hasParent to check...

**Comment by hsorby** [comment]
Maybe a little less of the grape juice Keri :)

**Comment by kerimoyle** [comment]
Fat fingers, small screen ...

**Comment by kerimoyle** [comment]
So if we're going to allow the component to be added to the "wrong" place, it will affect how we store the relationship between parents and children.  Since at present we can only have one parent, this is how the network gets messed up ... I think we have three options:
1.  Require children which are being added to have a NULL parent before they're added, and warn the user otherwise.  ie addComponent(component) will return an error when component already has any parent.
    - This will prevent some use cases but at least the error messages returned can be done at the appropriate time.  
    - It goes against allowing invalid model creation though.
    - We can keep the same storage classes as we have at the moment.
2.  Change the storage class to allow for more than one parent.
    - this will allow invalid models to be created, errors only at validation
    - the change will propagate through a whole lotta code ...
3.  Make sure that the child list of the ex-parent is updated when addComponent is called.
    - will get around the bug we currently have, but ...
    - does not address the issue of the user ending up with a network which depends on the order in which it was created.
    - validation errors may not be able to pinpoint where the error originated from ... only what it is.  

All in all, I'd probably vote for 1 as the least bad option ...?

**Comment by MichaelClerx** [comment]
Bit late to the game but I've found a nice way of working is to not have an `add_component(object)` method but only a `add_component(name)` instead, which creates the component and adds it to the model straight away. This is a gentle way of enforcing 1 (and means the component is created in a valid state, although the user is free to mess things up)

Big fan of allowing invalid models to be created (for example, if you need to modify a model in several steps its very awkward if every step needs to result in a valid model), but at the same time I would encourage an API design in which the obvious* way to do things results in valid models
(*whatever that means)

For what it's worth :-)
